daemon: Implement route-based device detection

joamaki opened the pull request:
This reimplements the device detection logic to use the route
information rather than detecting devices based on k8s nodeIP and
the device with the default route.

The devices are discovered by finding all devices mentioned in global
unicast routes and filtering out cilium-managed devices by prefix.

Motivation for this change is to handle cases like #15960 more reliably
and not have the user specify devices manually. This also prepares for
later work to detect devices at runtime to allow e.g. use of KPR with ENI,
which adds devices dynamically.

The reason for introducing DeviceManager is to later add support for
dynamically reconfiguring the datapath for devices added at runtime.
Also the detected devices are now also used for host firewall and
bandwidth manager so makes sense for this logic to be moved out.

Fixes: #15960

```release-note
Detect devices from global unicast routes in addition to only
looking for the device with the Kubernetes Node IP and the one with
default route.  This expands the set of devices used for kube-proxy
replacement, host firewall and bandwidth manager and should reduce
the need to specify devices manually.
```

joamaki committed:
daemon: Implement route-based device detection

This reimplements the device detection logic to use the route
information rather than detecting devices based on k8s nodeIP and
the device with the default route.

The devices are discovered by finding all devices mentioned in global
unicast routes and filtering out cilium-managed devices by prefix.

The reason for introducing DeviceManager is to later add support for
dynamically reconfiguring the datapath for devices added at runtime.
Also the detected devices are now also used for host firewall and
bandwidth manager so makes sense for this logic to be moved out.

Fixes: #15960

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,257 @@
+
+// updateDevicesFromRoutes processes a batch of routes and updates the set of
+// devices. Returns true if devices changed.
+func (dm *DeviceManager) updateDevicesFromRoutes(routes []netlink.Route) bool {
```
A follow-up PR will reuse this logic to detect devices at runtime and reload the datapath.

joamaki commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,257 @@
+	return nil
+}
+
+func (dm *DeviceManager) GetDevices() []string {
```
In a follow-up with runtime detection of devices the plan is to stop using `option.Config.Devices` throughout and rather call into device manager to ask for the list of devices as it may mutate over time. Hence I already have the state for the devices and this function here. 

See https://github.com/cilium/cilium/pull/17187 for the work in progress for the runtime detection and reasoning behind the structure of this file.

joamaki commented on daemon/cmd/daemon.go:
```diff
@@ -112,6 +112,8 @@ type Daemon struct {
 	monitorAgent *monitoragent.Agent
 	ciliumHealth *health.CiliumHealth
 
+	deviceManager *DeviceManager
```
A more suitable place for `DeviceManager` might be in `pkg/datapath`, but I'd like to leave it here for now.

joamaki commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,257 @@
+	}
+
+	detectDirectRoutingDev := option.Config.EnableNodePort && option.Config.DirectRoutingDevice == ""
+	detectIPv6MCastDev := option.Config.EnableIPv6NDP && len(option.Config.IPv6MCastDevice) == 0
```
```suggestion
	detectIPv6MCastDev := option.Config.EnableIPv6NDP && option.Config.IPv6MCastDevice == ""
```

borkmann commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,257 @@
+				// use it for direct routing and IPv6 multicast.
+				if addrs, err := netlink.AddrList(link, netlink.FAMILY_ALL); err == nil {
+					for _, a := range addrs {
+						if a.IP.Equal(nodeIP) {
```
Could we add the dev to the list of devs in any case if it matches k8s node ip (given here we know the intent for sure)?

joamaki replied:
Done.

joamaki commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,292 @@
+	if option.Config.EnableNodePort {
+		// Probe whether fast redirect is supported for L3 devices. This will
+		// invoke bpftool and requires root privileges.
+		l3DevOK = supportL3Dev()
```
Had to put the probing behind the EnableNodePort check as otherwise integration tests "Fatal()" when bpftool is missing (or if not run as root). Alternatively could make the probing not fatal on bpftool invocation failure and just return empty set of features, but not sure if that's any nicer.

brb replied:
The test is marked to be run as privileged. Should we just install bpftool in the env which runs the test suite?

joamaki replied:
Having bpftool installed is probably better. Will take a look at adding it.

joamaki replied:
Yep I think I meant the privileged unit tests. daemon_test.go I think was stumbling on this.

pchaigno replied:
AFAIK, `bpftool` should already be installed in all environments we use for integration tests. Did you mean unit tests maybe?

joamaki replied:
I think the solution as is is fine. The failing tests prior to moving call to supportL3Dev() here was the daemon_test.go which was non-privileged, so we wouldn't be able to query this with bpftool even if it was available.

brb commented on daemon/cmd/daemon.go:
```diff
@@ -603,7 +606,13 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
 	// which can be modified after the device detection.
-	handleNativeDevices(isKubeProxyReplacementStrict)
+	if err := d.deviceManager.Detect(); err != nil {
+		if isKubeProxyReplacementStrict {
```
@pchaigno We could extend this condition by appending `|| option.Config.EnableHostFirewall`?

pchaigno replied:
IIUC, we should extend this condition with all features that require devices to be set, so host firewall and bandwidth manager. IPsec as well if/once it uses this set of devices.

joamaki replied:
Done.

brb commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,296 @@
+)
+
+type DeviceManager struct {
+	mu      lock.Mutex
```
Q: can you embed the lock.Mutex instead of creating `mu`?

brb commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,296 @@
+
+type DeviceManager struct {
+	mu      lock.Mutex
+	devices map[string]bool
```
It's not clear what bool stands for. Maybe you could add the comment `// key meaning => val meaning`?

joamaki replied:
It's the suggested way of making a set in go: https://golang.org/doc/effective_go#maps. The boolean just marks existence.
I've seen "struct{}" being used elsewhere. Is there a reason to prefer that (I find it much uglier)?

pchaigno replied:
I personally like `map[string]void` but I don't think we have a precedent in the codebase. Regardless of what we use, could you add a comment explaining it's a set. It's not immediately obvious otherwise.

joamaki replied:
Will add, thanks!

joamaki replied:
I'll switch this to `struct{}` as that's the normal pattern.

brb replied:
> I've seen "struct{}" being used elsewhere

AFAIK `struct{}` should not allocate any mem for val, while `bool` does.

brb commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,296 @@
+	}
+}
+
+func (dm *DeviceManager) Close() {
```
What is supposed to be in the `Close` method?

joamaki replied:
This work was split off from PR #17187  and I wanted to get this part in first and with the basic structure in place already.
I can remove this if you want and add it later.

Eventually `Close()` will have something like this: https://github.com/cilium/cilium/pull/17187/files#diff-b7bf7344c04110932f57b676c0dce948271b6e2bb4a1587135e1cf5336958a5eR256.

pchaigno replied:
Nit: I'd prefer if it was removed until used. You never know what will happen with the second pull request; this method may not be needed anymore by the time it gets merged.

joamaki replied:
Yep cleaner. Removed.

brb commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,296 @@
+}
+
+// Detect tries to detect devices which are to be used for:
+// NodePort BPF, direct routing in NodePort BPF, IPsec and Bandwidth Manager.
```
I think you can merge `NodePort BPF` with `direct routing in NodePort BPF`. It's just an implementation detail.

pchaigno replied:
Host firewall is missing from the list. I wonder if we should just have a method `areDevicesRequired`, use it above, and refer to it here.

joamaki replied:
Good idea, added.

brb commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,296 @@
+
+	// Skip slave devices.
+	if link.Attrs().RawFlags&unix.IFF_SLAVE != 0 {
+		log.WithField(logfields.Device, name).Debugf("Skipping slave device")
```
Nit on terminology: s/slave/secondary device/?

pchaigno replied:
I think secondary as a different meaning in Linux terminology. I would also prefer to avoid this term here, but I'm not sure what to replace it with given it's specific to Linux. I ended up using it elsewhere in the codebase recently :disappointed:

joamaki replied:
I've not heard of "secondary device". We're testing whether the SLAVE flag is set, so we should talk about using the right terms? Or if we need to avoid that particular word we could use "lower" or "nested" device instead?

joamaki replied:
This is now "bonded device"

brb reviewed (commented):
LGTM, just a few nits.

pchaigno commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,296 @@
+	}
+
+	detectDevices := len(option.Config.Devices) == 0 &&
+		(option.Config.EnableNodePort || option.Config.EnableHostFirewall || option.Config.EnableBandwidthManager)
```
Plus `|| option.Config.EnableIPsec`?

joamaki replied:
Added.

joamaki replied:
Oops, only added it to areDevicesRequired(), but forgot to call it here. Fixed.

pchaigno commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,296 @@
+	}
+
+	changed := false
+links:
```
What is this label used for? I thought the typical use case was nested loops.

joamaki replied:
Ah heh, looks like refactoring at some point removed the nested loop, but I forgot this. I'll remove the label :)

pchaigno commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,296 @@
+		// e.g. skip loopback, multicast and link local routes.
+		if route.Dst != nil && !route.Dst.IP.IsGlobalUnicast() {
+			continue
+		}
```
This skip condition could be moved inside the second loop to avoid the `linkIndices` allocation and one iteration through routes.

joamaki replied:
Hmm I don't quite follow. We want to ignore all uninteresting routes here and not consider the links that don't have interesting routes.

pchaigno replied:
Yes, I'm not proposing to change that. But these four lines could be moved inside the second loop and it would achieve the same while avoiding the need for `linkIndices`, no?

joamaki replied:
I'm still not following how that would work. We may have hundreds of routes from which we want to create a set of interesting links to look at. The second loop doesn't deal with routes anymore so this check doesn't work there. Also please check the final code as the second commit modified this.

pchaigno replied:
> Also please check the final code as the second commit modified this.

Ah, forgot that. What I had in mind won't work with `linkInfo.hasDefaultRoute` :+1:

pchaigno commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,296 @@
+	for linkIndex := range linkIndices {
+		link, err := netlink.LinkByIndex(linkIndex)
+		if err != nil {
+			log.WithError(err).Warnf("Failed to get link by index %d", linkIndex)
```
```suggestion
			log.WithError(err).WithField(logfields.LinkIndex, linkIndex).Warn("Failed to get link by index")
```

pchaigno commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,296 @@
+
+	k8sNodeDev := ""
+	k8sNodeLink, err := findK8SNodeIPLink()
+	if err == nil {
```
Shouldn't we log the error if it is not nil?

joamaki replied:
No since k8s may not actually be always available in all use-cases. Though may be it should be here?

pchaigno replied:
Do you mean there's one error case that's expected? If that's the case, we should ignore that specific error and log the rest I think.

joamaki replied:
Well, it can fail in two ways: either the Node IP is not set (k8s not enabled, or something else horrible wrong), or the device can not be found that has the Node IP. Right now we only care about it failing to find the device if we need that information for the direct routing device or mcast device detection. 

I guess I see the point of failing more eagerly here. I'll think about what would be more exact thing to do here.

pchaigno replied:
If we're just worried about environment where K8s is not installed, we could simply check `k8s.IsEnabled()` before looking up `k8sNodeLink`.

joamaki replied:
Added the check and returning a better error now.

pchaigno commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,296 @@
+				option.Config.DirectRoutingDevice = dev
+				break
+			}
+		} else if k8sNodeLink != nil {
```
Maybe safer to do:
```suggestion
		} else if k8sNodeDev != "" {
```

pchaigno commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,296 @@
+		} else if k8sNodeLink != nil {
+			option.Config.DirectRoutingDevice = k8sNodeDev
+		} else {
+			return fmt.Errorf("Unable to determine direct routing device. Use --%s to specify them",
```
```suggestion
			return fmt.Errorf("Unable to determine direct routing device. Use --%s to specify it",
```

pchaigno commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,296 @@
+	}
+
+	// TODO(JM): Remove mutation of option.Config.Devices as it is unsafe. A follow up will add support
+	// for runtime detection of new devices and will switch cilium agent to use GetDevices() instead.
```
Nit: Maybe best to make a note of this in the issue used to track progress on runtime detection instead. It's the sort of comments we forget and tend to leave behind :grimacing:

joamaki replied:
Added TODO to roadmap issue and dropped the comment. I won't forget this one since it's one of the main blockers on #17187 :)

pchaigno commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,296 @@
+		l3DevOK = supportL3Dev()
+	}
+
+	detectDevices := len(option.Config.Devices) == 0 &&
```
If the device name expansion fails to find any matching devices, we will kick off the detection logic. Is that what we want? It may be a bit surprising for users if they make a mistake in the device pattern (not matching anything) and everything proceeds with unexpected devices (from detection).

joamaki replied:
Good catch. I think it's best to fail if expansion fails completely. I'll add that check. We do still want to allow expansion to partially not match so users can provide alternatives.

pchaigno commented on test/runtime/kafka.go:
```diff
@@ -134,11 +137,12 @@ var _ = Describe("RuntimeKafka", func() {
 
 	JustAfterEach(func() {
-		vm.ValidateNoErrorsInLogs(CurrentGinkgoTestDescription().Duration)
+		vm.ValidateNoErrorsInLogs(time.Since(testStartTime))
```
Could you send this fix as a separate pull request so we can backport it? :pray:

joamaki replied:
Yes will do!

joamaki replied:
https://github.com/cilium/cilium/pull/17351

pchaigno reviewed (changes requested):
LGTM overall, mostly nits below.

nbusseneau reviewed (approved):
Code changes LGTM.

Q: regarding CI status => I see Travis CI failing with
```
--- FAIL: Test (2.48s)
FAIL coverage: 3.8% of statements in ./...
FAIL	github.com/cilium/cilium/pkg/aws/eni	2.560s
FAIL make: *** [Makefile:205: integration-tests] Error 1
The command "./.travis/build.sh" exited with 2.
```
Did we investigate if this was related to the PR? It looks like #9604 at a first glance but just to be sure...

joamaki committed:
daemon: Do not detect devices for IPsec

Since IPsec is not yet using the detected devices, disable
detection if only IPsec is enabled.

This fixes a failure in IPsec L7 tests. Root cause still
unknown why detecting the devices causes issues.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

brb commented on daemon/cmd/devices.go:
```diff
@@ -185,13 +186,33 @@ func (dm *DeviceManager) isViableDevice(l3DevOK bool, link netlink.Link) bool {
+	// This is a workaround for kubernetes-in-docker. We want to avoid
+	// veth devices in general as they may be leftovers from another CNI.
+	if !hasDefaultRoute {
+		virtual, err := ethtool.IsVirtualDriver(name)
```
I'm wondering whether we really need to invoke the ethtool. AFAIK the netlink.Link should already embed the link type, so `_, virtual := netlink.(*Veth)` could do the job.

joamaki replied:
Thanks, that works nicely.

brb commented on daemon/cmd/devices.go:
```diff
@@ -255,22 +251,29 @@ links:
+	if len(option.Config.Devices) > 0 && len(expandedDevices) == 0 {
+		// User defined devices, but expansion yielded no devices. Fail here to not
+		// surprise with auto-detection.
+		return fmt.Errorf("Device wildcard expansion failed to detect devices. Please verify --%s option.",
```
I think we are changing the previous behavior (i.e., do the dev detection if the expansion has failed), which is fine, just that we need to add a note to the upgrade guide.

joamaki replied:
Added a note about that, plus about the change in device detection logic.

brb commented on daemon/cmd/devices.go:
```diff
@@ -100,36 +98,40 @@ func (dm *DeviceManager) Detect() error {
+		if err == nil {
+			k8sNodeDev = k8sNodeLink.Attrs().Name
+			dm.devices[k8sNodeDev] = struct{}{}
+		} else if k8s.IsEnabled() {
```
It might make sense to move the `k8s.IsEnabled()` check above the `detectDirectRoutingDev || detectIPv6MCastDev` check.

joamaki replied:
We only require k8s when direct routing device or ipv6 mcast device detection is necessary, so it doesn't make sense to move it there as then device detection won't work without k8s. Though do we ever need device detection without k8s? I'm not sure. If not, then we can just make this fail early when k8s is not enabled.

brb replied:
> Though do we ever need device detection without k8s?

Yes. In the `lb-only` mode when user specifies multiple devices, but doesn't specify the direct routing device. We should definitely keep the case `|devices| == 1 => dr_device = devices[0]` regardless of k8s presence.

brb commented on daemon/cmd/devices.go:
```diff
@@ -100,36 +98,40 @@ func (dm *DeviceManager) Detect() error {
-		} else {
-			return fmt.Errorf("Unable to determine direct routing device. Use --%s to specify it",
-				option.DirectRoutingDevice)
+			log.Infof("Detected %s=%s", option.DirectRoutingDevice, option.Config.DirectRoutingDevice)
```
While you are here, maybe you could fix the msg, so that we respect the structured logging?

joamaki replied:
Sure, good catch.

brb commented on daemon/cmd/devices.go:
```diff
@@ -100,36 +98,40 @@ func (dm *DeviceManager) Detect() error {
+		if detectIPv6MCastDev {
+			if k8sNodeLink != nil && k8sNodeLink.Attrs().Flags&net.FlagMulticast != 0 {
+				option.Config.IPv6MCastDevice = k8sNodeDev
+				log.Infof("Detected %s=%s", option.IPv6MCastDevice, option.Config.IPv6MCastDevice)
```
Ditto.

joamaki replied:
:+1:

brb reviewed (changes requested):
Cool, a few more comments, and we are getting ready to this PR to be merged.

BTW, regarding disabling the dev detection for the IPSec in the last commit - any clue why it's needed? /cc @pchaigno @jrfastab

pchaigno reviewed (changes requested):
Could you break commit `daemon: Address review comments on device detection` and merge it in the respective commits it fixes/improves? Having it in a separate commit hinders reviewing and doesn't make sense once merged.

qmonnet commented on daemon/cmd/devices_test.go:
```diff
@@ -7,7 +7,6 @@
 package cmd
 
 import (
-	"fmt"
```
Nit: this should be moved to the previous commit

joamaki replied:
Done.

pchaigno commented on daemon/cmd/daemon.go:
```diff
@@ -607,7 +610,14 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
 	// which can be modified after the device detection.
-	handleNativeDevices(isKubeProxyReplacementStrict)
+	if err := d.deviceManager.Detect(); err != nil {
+		if isKubeProxyReplacementStrict || areDevicesRequired() {
```
If `isKubeProxyReplacementStrict` is true, then `areDevicesRequired()` will always return true. So can we simplify this condition?

pchaigno commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,337 @@
+	}
+
+	l3DevOK := false
+	if option.Config.EnableNodePort {
```
Why `EnableNodePort`? From the comment below, I understood that we only need to restrict ourselves to L3 devices if fast redirect is used. So maybe we should check for the fast redirect config instead of the BPF NodePort one?

pchaigno commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,337 @@
+		return err
+	}
+
+	l3DevOK := false
```
Shouldn't this be initialized to true by default to allow such devices unless a feature forces us to exclude them?

pchaigno commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,337 @@
+		if detectIPv6MCastDev {
+			if k8sNodeLink != nil && k8sNodeLink.Attrs().Flags&net.FlagMulticast != 0 {
+				option.Config.IPv6MCastDevice = k8sNodeDev
+				log.WithField(option.IPv6MCastDevice, option.Config.IPv6MCastDevice).Infof("IPv6 multicast device detected")
```
Nit:
```suggestion
				log.WithField(option.IPv6MCastDevice, option.Config.IPv6MCastDevice).Info("IPv6 multicast device detected")
```

pchaigno commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,337 @@
+		}
+	}
+
+	// Skip bonding slave devices.
```
```suggestion
	// Skip lower bond devices.
```

pchaigno commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,337 @@
+	for _, p := range excludedDevicePrefixes {
+		if strings.HasPrefix(name, p) {
+			log.WithField(logfields.Device, name).
+				Warnf("Skipping device as it has excluded prefix '%s'", p)
```
Do we need to print a warning in this case? If I understand correctly, this means we'll print a warning per excluded device (if they are in routes, but most usually are) on each Cilium agent restart :grimacing: 

I think Debug would be good enough here.

pchaigno commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,337 @@
+
+	// Skip bonding slave devices.
+	if link.Attrs().RawFlags&unix.IFF_SLAVE != 0 {
+		log.WithField(logfields.Device, name).Warn("Skipping bonded device")
```
Here I think Debug or Info would be enough.

pchaigno commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,337 @@
+		_, virtual := link.(*netlink.Veth)
+		if virtual {
+			log.WithField(logfields.Device, name).
+				Warn("Ignoring veth device as it has no default route")
```
I think Debug would be enough here (it should be fairly obvious to users why we avoid veth devices).

pchaigno commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,337 @@
+	// Ignore L3 devices if we cannot support them.
+	if !l3DevOK && !mac.LinkHasMacAddr(link) {
+		log.WithField(logfields.Device, name).
+			Warn("Ignoring L3 device; >= 5.8 kernel is required.")
```
Info would be enough here I think (users might expect L3 devices to be detected, but it's not as if it necessarily needs immediate attention).

pchaigno commented on daemon/cmd/devices.go:
```diff
@@ -0,0 +1,337 @@
+}
+
+func areDevicesRequired() bool {
+	return len(option.Config.Devices) == 0 &&
```
This part of the condition feels a bit unexpected to me given the function's name. Whether devices are preconfigured or not shouldn't impact whether they are required. I would just remove it.

pchaigno reviewed (changes requested):
> > Could you break commit `daemon: Address review comments on device detection` and merge it in the respective commits it fixes/improves? Having it in a separate commit hinders reviewing and doesn't make sense once merged.
> 
> Ah I was just trying to make reviewing easier. I'll squash.

I didn't mean you should squash everything; that doesn't make reviews any easier ;-) 

If you're looking for guidance on how to split commits in pull requests, maybe a good starting point is https://kernelnewbies.org/PatchPhilosophy#How_to_break_up_changes. For this specific pull request, a good example is the code you moved from `daemon/cmd/kube_proxy_replacement{,_test}.go` to a new file; that move should be in its own commit, before the rest. It would help reviewers because we wouldn't have to re-review existing code that was just moved around.
